graph/articulationpoints.go

```diff
@@ -0,0 +1,53 @@
+// Algorithm to identify articulation points in a connected graph
+// You can read more about articulation points here: https://en.wikipedia.org/wiki/Biconnected_component
+// and the implementation and explanation here: https://cptalks.quora.com/Cut-Vertex-Articulation-point
+
+package graph
+
+import "github.com/TheAlgorithms/Go/math/min"
+
+func ArticulationPoint(graph Graph) []bool {
+	//initialize all the variables
+	time := 0                                         // time variable to keep track of the time of discovery_time of a vertex
+	is_ap := make([]bool, graph.vertices)             // boolean array to mark articulation points
+	visited := make([]bool, graph.vertices)           // boolean array to mark visited vertices
+	child_cnt := make([]int, graph.vertices)          // array to store the number of children of a vertex
+	discovery_time := make([]int, graph.vertices)     // array to store the time of discovery of a vertex
+	earliest_discovery := make([]int, graph.vertices) // array to store the earliest discovered vertex reachable from a vertex
+
+	articulationPointHelper(graph, 0, -1, &time, is_ap, visited, child_cnt, discovery_time, earliest_discovery)
+
+	if child_cnt[0] == 1 {
+		is_ap[0] = false // if the root has only one child, it is not an articulation point
+	}
+
+	return is_ap
+}
+
+func articulationPointHelper(graph Graph, vertex int, parent int, time *int, is_ap []bool, visited []bool, child_cnt []int, discovery_time []int, earliest_discovery []int) {
```

Comment:
This is actually very confusing haha.

Reply:
Are you comfortable with pointers? If so, I recommend creating a package scoped (unexported) struct (with the first letter lowercase) with these fields like so


```Go

type apHelper struct {
	graph              Graph
	vertex             int
	parent             int
	time               *int
	is_ap              []bool
	visited            []bool
	child_cnt          []int
	discovery_time     []int
	earliest_discovery []int
}
```

This will make it easier to recursively call the function if you change the definition to something like this

```Go
func apHelperFunc(ap *apHelper) {
	...
	// do the operations to the ap.graph or ap.parent or ap.visited etc
	// and it will be reflected to the main one.
}
```

With the above change you will also need to change the `ArticulationPoint` function to something like this

```Go
func ArticulationPoint(graph Graph) []bool {
	//initialize all the variables
	ap := apHelper{
		time:               0,
		is_ap:              make([]bool, graph.vertices),
		visited:            make([]bool, graph.vertices),
		child_cnt:          make([]int, graph.vertices),
		discovery_time:     make([]int, graph.vertices),
		earliest_discovery: make([]int, graph.vertices),
	}
	
	articulationPointHelper(&ap)
```

and that should do it.

That is IMO much cleaner than having it as arguments.

Let me know what you think.